README.md
<!-- PROJECT LOGO -->
<br />
<p align="center">
  <a href="https://www.mhh.de/">
    <img src="shiny/www/Logo_engl_schwarz.png" height="25%" width="25%" alt="MHH">
  </a>
  
  <a href="https://www.mhh.de/genomics">
    <img src="shiny/www/RCUG_Logo.png" height="15%" width="15%" alt="Logo">
  </a>
  

  <h2 align="center">scrnaseq_app</h2>

  <p align="center">
    This is an R/Shiny app for visualisation of single cell RNA-seq (scRNASeq) data. Based on a provided Seurat object, one or more genes can be selected through a user interface, and several plots are generated, including feature plots, ridge plots, violin plots and dot plots. This project is based on <a href="https://github.com/ktrns/scrnaseq">scrnaseq</a>. The scrnaseq workflow was and is being developed by <a href="https://github.com/ktrns">Katrin Sameith</a> and <a href="https://github.com/andpet0101">Andreas Petzold</a> at the <a href="https://genomecenter.tu-dresden.de/about-us">DRESDEN-concept Genome Center (Dresden, Germany)</a>, in collaboration with the <a href="https://www.mhh.de/genomics">Research Core Unit Genomics (Hannover, Germany)</a>.
    <br />
    <a href="https://github.com/MHH-RCUG/scrnaseq_app"><strong>Explore the docs »</strong></a>
    <br />
    <br />
     :bug:
    <a href="https://github.com/MHH-RCUG/scrnaseq_app/issues">Report Bug</a>
    ·  :question:
    <a href="https://github.com/MHH-RCUG/scrnaseq_app/issues">Request Feature</a>
  </p>
</p>



<!-- TABLE OF CONTENTS -->
## Table of Contents

* [About the Project](#about-the-project)
  * [Built With](#built-with)
* [Usage](#usage)
  * [Upload .rds file](#upload-rds-file)
  * [Select genes](#select-genes)
* [Getting Started](#getting-started)
* [Contributing](#contributing)
* [Contact](#contact)
* [Acknowledgements](#acknowledgements)



<!-- ABOUT THE PROJECT -->
## About The Project

![screenshot](https://github.com/MHH-RCUG/scrnaseq_app/blob/master/www/Screenshot%202020-10-02%20220712.png)
This app is dependent on data generated by Seurat, which must be run first. The app reads an .rds file which contains the Seurat object and all associated data. 
The user can then select genes through a list or an Excel file (.xlsx). The expression of these genes is then displayed in a) UMAP plots b) Ridge plots c) Violin plots d) Dot plots or e) a global Heatmap. The generated plots can then be downloaded in .pdf and .png format.

### Built With

* [R](https://www.r-project.org/)
* [Shiny from RStudio](https://shiny.rstudio.com/)
* [Seurat](https://satijalab.org/seurat/)



<!-- USAGE EXAMPLES -->
## Usage

To get the app running, you need to follow these simple steps.

### Upload .rds file
First, you need to upload an .rds file containing your Seurat object. If you do not have this file yet but would still like to try the app, you can use this [example](https://owncloud.gwdg.de/index.php/s/rRawkhIOVe1T5qi).

### Select genes

#### Select genes through the user interface
Once the .rds file is uploaded and processed, you can click on the field below `Genes` and select your genes by clicking or searching. To delete genes again, you have to click on the gene and hit the backspace key.

#### Select genes through an Excel file with header
You can select genes by uploading an .xlsx file ([example](https://owncloud.gwdg.de/index.php/s/ZwY0iVPji6uBVKO)).  
The spreadsheet has to look as follows:

| Ensembl_IDs | Gene_Names |
| :--- | :--- |
| First ID  | First Name |
| Second ID | Second Name |

The app will only read the first column including the Ensembl IDs. The second column `Genes` is optional.

#### Select genes through an Excel file without header
If your .xlsx file contains no header, please uncheck the checkbox. Otherwise, the first Ensembl ID will not show up in the selection.

### Aspect Ratio
In order to change the aspect ratio of the plots you can change the pixels of the x and y-axis. By clicking on `Default settings for axes` the default settings will be restored: `X = 1024px`and `Y = 576px`

### Download
Once the plots are generated, you can download them. The download is an archive which contains a **PDF and PNG** version of the selected plots.
If you want to rename your archive before downloading, you can do so in the given field.

### Troubleshooting

  * Should the app hang or stop responding, you can reset it by reloading the data (RDS file) into the app.

<!-- GETTING STARTED -->
## Getting Started

### Installation

#### Singularity
If you are within the network of the MHH, use this [link](http://172.24.148.210:3838/) if the singularity container is running.

#### Local
To run a copy of this app locally, follow these steps:
1. Go to [Releases](https://github.com/MHH-RCUG/scrnaseq_app/releases) and download the latest release
2. Unzip the project und run scrnaseq_app.R with [RStudio](https://rstudio.com/)
3. Install required libraries


<!-- CONTRIBUTING -->
## Contributing

Contributions are what make the open source community such an amazing place to be learn, inspire, and create. Any contributions you make are **greatly appreciated**.

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request



<!-- LICENSE 
## License

Distributed under the MIT License. See `LICENSE` for more information.
-->


<!-- CONTACT -->
## Contact

Project Link: [https://github.com/MHH-RCUG/scrnaseq_app](https://github.com/MHH-RCUG/scrnaseq_app)



<!-- ACKNOWLEDGEMENTS -->
## Acknowledgements

* [scrnaseq](https://github.com/ktrns/scrnaseq)
* [README-Template](https://github.com/othneildrew/Best-README-Template)  


[![Contributors](https://img.shields.io/github/contributors/MHH-RCUG/scrnaseq_app)](https://github.com/MHH-RCUG/scrnaseq_app/graphs/contributors)
[![GitHub issues](https://img.shields.io/github/issues/MHH-RCUG/scrnaseq_app)](https://github.com/MHH-RCUG/scrnaseq_app/issues)
[![GitHub forks](https://img.shields.io/github/forks/MHH-RCUG/scrnaseq_app)](https://github.com/MHH-RCUG/scrnaseq_app/network)
[![GitHub stars](https://img.shields.io/github/stars/MHH-RCUG/scrnaseq_app)](https://github.com/MHH-RCUG/scrnaseq_app/stargazers)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0] - 02.01.2021

### Added

- Switched to [shinydashboard](http://rstudio.github.io/shinydashboard/)
- Option to change resoultion of plots without changing the size/ pixel

### Changed

- UI to be more guided 

### Removed

- nothing


## [1.0.0] - 11.11.2020
### First release
Changes will be documtented like this:
```
## [version] - dd.mm.yyyy

### Added

- x

### Changed

- y

### Removed

- z
``` 

## [0.0.1] - 02-09-2020

### Added

- Created a [CHANGELOG](https://github.com/MHH-RCUG/scrnaseq_app/blob/master/CHANGELOG.md) to keep track of changes

### Changed

- y

### Removed

- z


[unreleased]: https://github.com/MHH-RCUG/scrnaseq_app/compare/dev
[0.0.1]: https://github.com/MHH-RCUG/scrnaseq_app/releases/tag/0.0.1
[1.0.0]: https://github.com/MHH-RCUG/scrnaseq_app/releases/tag/1.0.0
[1.1.0]: https://github.com/MHH-RCUG/scrnaseq_app/releases/tag/1.1.0


app.R
# Author: Marius Rueve

# ! This is the user-interface definition of a Shiny web application. You can
# run the application by clicking 'Run App' above.

# ! For convenience click "Run External" to run in you default browser

# Global --------------------
# Load packages
library(shiny)
library(Seurat)
library(ggplot2)
library(shinythemes)
library(markdown)
library(readxl)
library(shinyalert)
library(shinyjs)
library(zip)

# Set max upload size to 300 MB
options(shiny.maxRequestSize = 5000 * 1024 ^ 2)

# Parameter list for convenience
param = list()
param$col = "palevioletred"

# Load data if upload is broken
#sc = readRDS(file = "pbmc_2020-06-09.rds")
#features_names_ids = paste(rownames(sc[["RNA"]][[]]), "_", sc[["RNA"]][[]][,1], sep = "")

# Empty object so Input UI gets rendered
features_names_ids = NULL

# Define global variable to store plots
stored_FeaturePlots = NULL
stored_RidgePlotRaws = NULL
stored_RidgePlotNorms = NULL
stored_ViolinPlotRaws = NULL
stored_ViolinPlotNorms = NULL
stored_DotPlot = NULL
stored_Heatmap = NULL

# UI --------------------
ui = fluidPage(

  #theme = shinytheme("paper"),

  # Place RCUG logo at top right
  img(
    src = "RCUG_Logo.png",
    height = "10%",
    width = "10%",
    align = "right"
  ),

  # Place MHH logo at top right
  img(
    src = "Logo_engl_schwarz.png",
    height = "15%",
    width = "15%",
    align = "right"
  ),

  # Application title
  titlePanel("scrnaseq_app"),

  # Sidebar
  sidebarPanel(
    width = 3,
    h4("1. Upload:"),

    # Upload of .rds file
    fileInput(
      inputId = "rds_file",
      label = "Upload Seurat file: (.rds)",
      accept = ".rds",
      buttonLabel = "Browse",
      placeholder = "Please upload .rds file!"
    ),

    # UI will be rendered once the .rds file has been uploaded
    uiOutput("insert_ui"),
  ),

  # Main panel
  mainPanel(

    useShinyjs(),
    useShinyalert(),

    # Set tabs
    tabsetPanel(
      type = "tabs",
      tabPanel("FeaturePlot", uiOutput("ui_feature")),
      tabPanel("RidgePlot (Raw)", uiOutput("ui_ridge_raw")),
      tabPanel("RidgePlot (Norm)", uiOutput("ui_ridge_norm")),
      tabPanel("ViolinPlot (Raw)", uiOutput("ui_vln_raw")),
      tabPanel("ViolinPlot (Norm)", uiOutput("ui_vln_norm")),
      tabPanel("DotPlot", plotOutput("plot_dotplot")),
      tabPanel("Heatmap", plotOutput("plot_heatmap")),
      tabPanel("Help", includeMarkdown("HELP.md"))
    )
  )
)


# Server --------------------
server = function(input, output, session) {

  output$insert_ui = renderUI({
    req(input$rds_file)
    tagList(
      tags$hr(),
      h4("2. Select genes:"),
      selectInput(
        "genes",
        "Select through list:",
        features_names_ids,
        multiple = TRUE
      ),

      # Select genes
      actionButton("clear_selection", "Clear selection"),
      fileInput(
        inputId = "xlsx_file",
        label = "Select through excel file: (.xlsx)",
        accept = ".xlsx",
        buttonLabel = "Browse",
        placeholder = "No file selected"
      ),

      tags$hr(),

      fluidRow(
        column(
          width = 6,
          numericInput(
            inputId = "x_axis",
            label = "X-Axis (px):",
            value = 1280,
            min = 1,
            max = 3000
          )
        ),

        column(
          width = 6,
          ofset = 3,
          numericInput(
            inputId = "y_axis",
            label = "Y-Axis (px):",
            value = 720,
            min = 1,
            max = 3000
          )
        )
      ),
      
      actionButton(inputId = "preset_512",
                   label = "512x288 px"),
      
      actionButton(inputId = "preset_1024",
                   label = "1024x576 px"),

      actionButton(inputId = "restore_axes",
                   label = "Default settings"),

      tags$hr(),

      h4("3. Download:"),

      textInput(
        inputId = "archive_download",
        label = "Enter name of archive (.zip):",
        value = paste0("Download", "_", Sys.Date())
      ),

      fluidRow(
        column(
          5,
          checkboxInput("check_featureplot", "FeaturePlot", value = TRUE)
        ),
        column(
          5,
          checkboxInput("check_ridgeplot_raw", "RidgePlot (Raw)", value = TRUE)
        ),
        column(
          5,
          checkboxInput("check_ridgeplot_norm", "RidgePlot (Norm)", value = TRUE)
        ),
        column(
          5,
          checkboxInput("check_vlnplot_raw", "ViolinPlot (Raw)", value = TRUE)
        ),
        column(
          5,
          checkboxInput("check_vlnplot_norm", "ViolinPlot (Norm)", value = TRUE)
        ),
        column(5, checkboxInput("check_dotplot", "DotPlot", value = TRUE))
      ),

      downloadButton("download_plots")
    )
  })

  # Upload ====================
  # .rds file upload
  sc = reactive({
    inFile = input$rds_file
    if (is.null(inFile)) {
      tmp = NULL
    } else {
      shinyalert(
        title = "Please wait!",
        text = "Upload complete! Please wait while the data is being processed!",
        size = "s",
        type = "info",
        showConfirmButton = FALSE
      )
      tmp = readRDS(inFile$datapath)
    }
    return(tmp)
  })

  observeEvent(sc(), {
    if (!is.null(sc())) {
      features_names_ids <<-
        paste(rownames(sc()[["RNA"]][[]]), "_", sc()[["RNA"]][[]][, 1], sep = "")
      #print(features_names_ids)
      updateSelectInput(session, "genes", "Select Genes:", features_names_ids)
    }
    shinyjs::delay(500, shinyjs::runjs("swal.close();"))
  })

  # Excel file upload
  excel_genes = reactive({
    inFile = input$xlsx_file
    if (is.null(inFile)) {
      tmp = NULL
    } else {
      tmp = readxl::read_excel(
        path = inFile$datapath,
        sheet = 1,
        range = cell_cols("A:B"),
        col_names = FALSE
      )
      shinyalert(
        title = "Please wait!",
        text = "Upload complete! Please wait while the data is being processed!",
        size = "s",
        type = "info",
        showConfirmButton = FALSE
      )
    }
    return(tmp)
  })

  # Observe event: when excel file has been uploaded and read
  observeEvent(excel_genes(), {
    if (length(excel_genes()) == 0) {
      shinyjs::runjs("swal.close();")
      shinyjs::delay(1000)
      shinyalert(
        title = "Error!",
        text = "The Excel file was empty!",
        size = "s",
        type = "error",
        showConfirmButton = TRUE
      )
      return(NULL)
    } else{
      excel_list = unlist(excel_genes()[, 1])
      x = features_names_ids[unlist(lapply(excel_list, function(one_gene)
        grep(one_gene, features_names_ids)))]
      updateSelectInput(
        session = session,
        inputId = "genes",
        label = "Select Genes:",
        choices = features_names_ids,
        selected = x
      )
      shinyjs::delay(500, shinyjs::runjs("swal.close();"))
    }
  })

  # If Button "clear selection" is pressed the selection is null
  observeEvent(input$clear_selection, {
    updateSelectInput(session, "genes", "Select Genes:", features_names_ids)
  })
  
  observeEvent(input$preset_1024, {
    updateNumericInput(
      session,
      "x_axis",
      "X-Axis (px):",
      value = 1024,
      min = 1,
      max = 3000
    )
    updateNumericInput(
      session,
      "y_axis",
      "Y-Axis (px):",
      value = 576,
      min = 1,
      max = 3000
    )
  })
  
  observeEvent(input$preset_512, {
    updateNumericInput(
      session,
      "x_axis",
      "X-Axis (px):",
      value = 512,
      min = 1,
      max = 3000
    )
    updateNumericInput(
      session,
      "y_axis",
      "Y-Axis (px):",
      value = 288,
      min = 1,
      max = 3000
    )
  })

  # Button to restore default settings for axes
  observeEvent(input$restore_axes, {
    updateNumericInput(
      session,
      "x_axis",
      "X-Axis (px):",
      value = 1280,
      min = 1,
      max = 3000
    )
    updateNumericInput(
      session,
      "y_axis",
      "Y-Axis (px):",
      value = 720,
      min = 1,
      max = 3000
    )
  })


  # Plots =================================

  # Rendering Plots, location of Seurat functions
  observeEvent(input$genes, {
    for (i in 1:length(input$genes)) {
      # Feature Plots
      stored_FeaturePlots[[i]] <<- # <<- for global assignments
        Seurat::FeaturePlot(
          sc(),
          features = unlist(strsplit(input$genes[[i]], "_"))[c(T, F)],
          cols = c("lightgrey", param$col),
          label = TRUE
        ) +
        theme_light() + theme(panel.border = element_blank())

      # Ridge Plots Raw
      stored_RidgePlotRaws[[i]] <<-
        Seurat::RidgePlot(
          sc(),
          features = unlist(strsplit(input$genes[[i]], "_"))[c(T, F)],
          assay = "RNA",
          slot = "counts"
        ) +
        theme_light() + theme(panel.border = element_blank()) +
        labs(color = "Cell identity", fill = "Cell identity") +
        ylab("Cluster")

      # Ridge Plots Norm
      stored_RidgePlotNorms[[i]] <<-
        Seurat::RidgePlot(sc(),
                          features = unlist(strsplit(input$genes[[i]], "_"))[c(T, F)],
                          slot = "data") +
        theme_light() + theme(panel.border = element_blank()) +
        labs(color = "Cell identity", fill = "Cell identity") +
        ylab("Cluster")

      # Violin Plots Raw
      stored_ViolinPlotRaws[[i]] <<-
        Seurat::VlnPlot(
          sc(),
          features = unlist(strsplit(input$genes[[i]], "_"))[c(T, F)],
          assay = "RNA",
          slot = "counts",
          pt.size = 0.2
        ) +
        theme_light() + theme(panel.border = element_blank()) +
        labs(color = "Cell identity", fill = "Cell identity") +
        xlab("Cluster")

      # Violin Plots Norm
      stored_ViolinPlotNorms[[i]] <<-
        Seurat::VlnPlot(sc(),
                        features = unlist(strsplit(input$genes[[i]], "_"))[c(T, F)],
                        pt.size = 0.2) +
        theme_light() + theme(panel.border = element_blank()) +
        labs(color = "Cell identity", fill = "Cell identity") +
        xlab("Cluster")

      # DotPlot
      stored_DotPlot <<-
        Seurat::DotPlot(
          sc(),
          features = unlist(strsplit(input$genes, "_"))[c(T, F)],
          cols = c("lightgrey", param$col)
        ) +
        theme_light() + theme(panel.border = element_blank()) +
        ylab("Cluster") +
        theme(axis.text.x = element_text(
          angle = 90,
          hjust = 1,
          vjust = .5
        ))

      tryCatch({
        # Heatmap
        stored_Heatmap <<-
          Seurat::DoHeatmap(
            sc(),
            features = unlist(strsplit(input$genes, "_"))[c(T, F)],
            group.colors = param$col_clusters,
            slot = "counts"
          ) +
          NoLegend()
      },
      warning = function(w){
          showNotification(
            ui = "One or more of the selected genes could not be found in the 
                  default search locations. These gene(s) will not show up in the Heatmap!",
            duration = NULL,
            id = "heatmap_warning",
            type = "warning"
          )
      }
      )
    }
  })


  # UI FeaturePlot #############################
  output$ui_feature = renderUI({
    tmp = list()

    if (length(input$genes) == 0) {
      return(NULL)
    }
    for (i in 1:length(input$genes)) {
      tmp[[i]] =  plotOutput(
        outputId = paste0("plot_feature", i),
        width = paste0(input$x_axis, "px"),
        height = paste0(input$y_axis, "px")
      )
    }
    return(tmp)
  })


  # UI RidgePlot Raw #############################
  output$ui_ridge_raw = renderUI({
    tmp = list()

    if (length(input$genes) == 0) {
      return(NULL)
    }
    for (i in 1:length(input$genes)) {
      tmp[[i]] =  plotOutput(
        outputId = paste0("plot_ridge_raw", i),
        width = paste0(input$x_axis, "px"),
        height = paste0(input$y_axis, "px")
      )
    }
    return(tmp)
  })


  # UI RidgePlot Normalised #############################
  output$ui_ridge_norm = renderUI({
    tmp = list()

    if (length(input$genes) == 0) {
      return(NULL)
    }
    for (i in 1:length(input$genes)) {
      tmp[[i]] =  plotOutput(
        outputId = paste0("plot_ridge_norm", i),
        width = paste0(input$x_axis, "px"),
        height = paste0(input$y_axis, "px")
      )
    }
    return(tmp)
  })


  # UI ViolinPlot Raw #############################
  output$ui_vln_raw = renderUI({
    tmp = list()

    if (length(input$genes) == 0) {
      return(NULL)
    }
    for (i in 1:length(input$genes)) {
      tmp[[i]] =  plotOutput(
        outputId = paste0("plot_vln_raw", i),
        width = paste0(input$x_axis, "px"),
        height = paste0(input$y_axis, "px")
      )
    }
    return(tmp)
  })


  # UI ViolinPlot Normalised #############################
  output$ui_vln_norm = renderUI({
    tmp = list()

    if (length(input$genes) == 0) {
      return(NULL)
    }
    for (i in 1:length(input$genes)) {
      tmp[[i]] =  plotOutput(
        outputId = paste0("plot_vln_norm", i),
        width = paste0(input$x_axis, "px"),
        height = paste0(input$y_axis, "px")
      )
    }
    return(tmp)
  })


  # renderPlots() --> plotOutput() --> renderUI()
  observe({
    for (i in 1:length(input$genes)) {
      local({
        # because expressions are evaluated at app init
        ii = i

        # This chunks takes the stored plots from the global variable and renders a reactive plots that is
        # suitable for assigning to an `output` slot; passing vector to UI
        output[[paste0("plot_feature", ii)]] = renderPlot(stored_FeaturePlots[[ii]])

        output[[paste0("plot_ridge_raw", ii)]] = renderPlot(stored_RidgePlotRaws[[ii]])

        output[[paste0("plot_ridge_norm", ii)]] = renderPlot(stored_RidgePlotNorms[[ii]])

        output[[paste0("plot_vln_raw", ii)]] = renderPlot(stored_ViolinPlotRaws[[ii]])

        output[[paste0("plot_vln_norm", ii)]] = renderPlot(stored_ViolinPlotNorms[[ii]])

        output$plot_dotplot = renderPlot(stored_DotPlot,
                                         width = input$x_axis,
                                         height = input$y_axis)

        output$plot_heatmap = renderPlot(stored_Heatmap,
                                         width = input$x_axis,
                                         height = input$y_axis)

      })
    }
  })


  # Downloads =================================
  output$download_plots = downloadHandler(
    filename = function() {
      # Takes the filename from the input of the user
      paste0(input$archive_download, ".zip")
    },
    content = function(file) {
      #temporarily switch to the temp dir, in case you do not have write permission to the current working directory
      owd = setwd(tempdir())
      on.exit(setwd(owd))

      # Idea to check if all devices are off on exit
      # on.exit({
      #  setwd(owd)
      #  while length(dev.list())>1; do
      #    dev.off()
      # })

      files = NULL
      on.exit(unlink(files))

      if (length(input$genes) == 0) {
        # Message if download fails due to no genese selected
        shinyalert(
          title = "Download failed!",
          text = "Failed to download plots! Please make sure to upload .rds file and select genes before downloading!",
          size = "s",
          type = "error"
        )
        return(NULL)
      } else{
        # Message that the archive is being created and the download will start soon
        shinyalert(
          title = "Please wait!",
          text = "The creation of files (.png and .pdf) can take a while. The download will start, once everything is ready.",
          size = "s",
          type = "info",
          showConfirmButton = FALSE
        )
      }


      # Download FeaturPlot #############################
      if (input$check_featureplot == TRUE) {
        # PNG
        for (i in 1:length(input$genes)) {
          fileName_png = paste0("FeaturePlot_", input$genes[[i]], ".png")

          png(fileName_png,
              width = input$x_axis,
              height = input$y_axis)
          print(stored_FeaturePlots[[i]])
          dev.off()
          files = c(fileName_png, files)
        }
        # PDF
        fileName_pdf = "FeaturePlot.pdf"
        pdf(
          file = fileName_pdf,
          width = (input$x_axis / 96),
          height = (input$y_axis / 96)
        )
        for (i in 1:length(input$genes)) {
          print(stored_FeaturePlots[[i]])
        }
        dev.off()
        files = c(fileName_pdf, files)
      }


      # Download RidgePlot Raw #############################
      if (input$check_ridgeplot_raw == TRUE) {
        # PNG
        for (i in 1:length(input$genes)) {
          fileName_png = paste0("RidgePlot_Raw_", input$genes[[i]], ".png")

          png(fileName_png,
              width = input$x_axis,
              height = input$y_axis)
          print(stored_RidgePlotRaws[[i]])
          dev.off()
          files = c(fileName_png, files)
        }
        # PDF
        fileName_pdf = "RidgePlot_Raw.pdf"
        pdf(
          file = fileName_pdf,
          width = (input$x_axis / 96),
          height = (input$y_axis / 96)
        )
        for (i in 1:length(input$genes)) {
          print(stored_RidgePlotRaws[[i]])
        }
        dev.off()
        files = c(fileName_pdf, files)
      }


      # Download RidgePlot Normalised #############################
      if (input$check_ridgeplot_norm == TRUE) {
        # PNG
        for (i in 1:length(input$genes)) {
          fileName_png = paste0("RidgePlot_Norm_", input$genes[[i]], ".png")

          png(fileName_png,
              width = input$x_axis,
              height = input$y_axis)
          print(stored_RidgePlotNorms[[i]])
          dev.off()
          files = c(fileName_png, files)
        }
        # PDF
        fileName_pdf = "RidgePlot_Norm.pdf"
        pdf(
          file = fileName_pdf,
          width = (input$x_axis / 96),
          height = (input$y_axis / 96)
        )
        for (i in 1:length(input$genes)) {
          print(stored_RidgePlotNorms[[i]])
        }
        dev.off()
        files = c(fileName_pdf, files)
      }


      # Download ViolinPlot Raw #############################
      if (input$check_vlnplot_raw == TRUE) {
        # PNG ViolinPlot Raw
        for (i in 1:length(input$genes)) {
          fileName_png = paste0("ViolinPlot_Raw_", input$genes[[i]], ".png")

          png(fileName_png,
              width = input$x_axis,
              height = input$y_axis)
          print(stored_ViolinPlotRaws[[i]])
          dev.off()
          files = c(fileName_png, files)
        }
        # PDF ViolinPlot Raw
        fileName_pdf = "ViolinPlot_Raw.pdf"
        pdf(
          file = fileName_pdf,
          width = (input$x_axis / 96),
          height = (input$y_axis / 96)
        )
        for (i in 1:length(input$genes)) {
          print(stored_ViolinPlotRaws[[i]])
        }
        dev.off()
        files = c(fileName_pdf, files)
      }


      # Download ViolinPlot Normalised #############################
      if (input$check_vlnplot_norm == TRUE) {
        # PNG ViolinPlot Normalised
        for (i in 1:length(input$genes)) {
          fileName_png = paste0("ViolinPlot_Norm_", input$genes[[i]], ".png")

          png(fileName_png,
              width = input$x_axis,
              height = input$y_axis)
          print(stored_ViolinPlotNorms[[i]])
          dev.off()
          files = c(fileName_png, files)
        }
        # PDF ViolinPlot Normalised
        fileName_pdf = "ViolinPlot_Norm.pdf"
        pdf(
          file = fileName_pdf,
          width = (input$x_axis / 96),
          height = (input$y_axis / 96)
        )
        for (i in 1:length(input$genes)) {
          print(stored_ViolinPlotNorms[[i]])
        }
        dev.off()
        files = c(fileName_pdf, files)
      }


      # Download DotPlot #############################
      if (input$check_dotplot == TRUE) {
        # PNG
        fileName_png = "DotPlot.png"

        png(fileName_png,
            width = input$x_axis,
            height = input$y_axis)
        print(stored_DotPlot)
        dev.off()
        files = c(fileName_png, files)
        # PDF
        fileName_pdf = "DotPlot.pdf"
        pdf(
          file = fileName_pdf,
          width = (input$x_axis / 96),
          height = (input$y_axis / 96)
        )
        print(stored_DotPlot)
        dev.off()
        files = c(fileName_pdf, files)
      }


      # Create zip file for Download, uses array of files
      zip(zipfile = file, files =  files)
      shinyjs::runjs("swal.close();")
    },
    contentType = "application/zip"
  )
}

# Run the application
shinyApp(ui = ui, server = server)
